### PR TITLE
[air] update DummyTrainer to handle DatasetPipelines

### PR DESCRIPTION
**Pull request by matthewdeng**
## Why are these changes needed?

1. Update `DummyTrainer` to take `num_epochs` instead of `runtime_seconds`.
    1. Ray Train expects equal number of calls to `train.report()`. Different workers may run at different speeds and terminate after different epoch numbers, which causes an error.
2. Add `generate_epochs` to support `DatasetPipeline` when `use_stream_api` is True.
3. Update `__main__` code to support testing different configurations.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/ray-air/doc_code/air_ingest.py [hunk 1]
```diff
@@ -25,7 +25,7 @@
     scaling_config={"num_workers": 1, "use_gpu": False},
     datasets={"train": dataset},
     preprocessor=preprocessor,
-    runtime_seconds=1,  # Stop after this amount or time or 1 epoch is read.
+    num_epochs=1,  # Stop after this number of epochs is read.
```
Comment (ericl) [inline doc/source/ray-air/doc_code/air_ingest.py:28]: Can you check this runs in a reasonable amount of time? The dataset is pretty big above.

**Review by ericl — APPROVED** [review]
Looks great!

File: doc/source/ray-air/doc_code/air_ingest.py (hunk 1 above)
Comment (matthewdeng) [reply doc/source/ray-air/doc_code/air_ingest.py:28]: Yep! (This should be the exact same as before change)

```
(DummyTrainer pid=38120) Preprocessed datasets in 45.017583959 seconds
(DummyTrainer pid=38120) Preprocessor Chain(preprocessors=(BatchMapper(fn=<lambda>), BatchMapper(fn=<lambda>)))
(DummyTrainer pid=38120) Preprocessor transform stats:
(DummyTrainer pid=38120) 
(DummyTrainer pid=38120) Stage 1 read->map_batches: 100/100 blocks executed in 13.09s
(DummyTrainer pid=38120) * Remote wall time: 106.0ms min, 567.14ms max, 334.9ms mean, 33.49s total
(DummyTrainer pid=38120) * Remote cpu time: 104.79ms min, 372.48ms max, 232.06ms mean, 23.21s total
(DummyTrainer pid=38120) * Peak heap memory usage (MiB): 440090624000.0 min, 1669951488000.0 max, 1071603097600 mean
(DummyTrainer pid=38120) * Output num rows: 500 min, 500 max, 500 mean, 50000 total
(DummyTrainer pid=38120) * Output size bytes: 102400128 min, 102400128 max, 102400128 mean, 10240012800 total
(DummyTrainer pid=38120) * Tasks per node: 100 min, 100 max, 100 mean; 1 nodes used
(DummyTrainer pid=38120) 
(DummyTrainer pid=38120) Stage 2 map_batches: 100/100 blocks executed in 31.84s
(DummyTrainer pid=38120) * Remote wall time: 60.24ms min, 1.31s max, 306.96ms mean, 30.7s total
(DummyTrainer pid=38120) * Remote cpu time: 53.86ms min, 257.69ms max, 133.78ms mean, 13.38s total
(DummyTrainer pid=38120) * Peak heap memory usage (MiB): 1465139200000.0 min, 2456166400000.0 max, 2040879800320 mean
(DummyTrainer pid=38120) * Output num rows: 500 min, 500 max, 500 mean, 50000 total
(DummyTrainer pid=38120) * Output size bytes: 102400128 min, 102400128 max, 102400128 mean, 10240012800 total
(DummyTrainer pid=38120) * Tasks per node: 100 min, 100 max, 100 mean; 1 nodes used
(DummyTrainer pid=38120) 
```
```
(BaseWorkerMixin pid=38238) Time to read all data 6.414166145 seconds
(BaseWorkerMixin pid=38238) P50/P95/Max batch delay (s) 0.046238997000000115 0.13754768315000004 0.49342400799999986
(BaseWorkerMixin pid=38238) Num epochs read 1
(BaseWorkerMixin pid=38238) Num batches read 100
(BaseWorkerMixin pid=38238) Num bytes read 9765.62 MiB
(BaseWorkerMixin pid=38238) Mean throughput 1522.51 MiB/s
(BaseWorkerMixin pid=38238) Ingest stats from rank=0:
(BaseWorkerMixin pid=38238) 
(BaseWorkerMixin pid=38238) Stage 1 read->map_batches: 100/100 blocks executed in 13.09s
(BaseWorkerMixin pid=38238) * Remote wall time: 106.0ms min, 567.14ms max, 334.9ms mean, 33.49s total
(BaseWorkerMixin pid=38238) * Remote cpu time: 104.79ms min, 372.48ms max, 232.06ms mean, 23.21s total
(BaseWorkerMixin pid=38238) * Peak heap memory usage (MiB): 440090624000.0 min, 1669951488000.0 max, 1071603097600 mean
(BaseWorkerMixin pid=38238) * Output num rows: 500 min, 500 max, 500 mean, 50000 total
(BaseWorkerMixin pid=38238) * Output size bytes: 102400128 min, 102400128 max, 102400128 mean, 10240012800 total
(BaseWorkerMixin pid=38238) * Tasks per node: 100 min, 100 max, 100 mean; 1 nodes used
(BaseWorkerMixin pid=38238) 
(BaseWorkerMixin pid=38238) Stage 2 map_batches: 100/100 blocks executed in 31.84s
(BaseWorkerMixin pid=38238) * Remote wall time: 60.24ms min, 1.31s max, 306.96ms mean, 30.7s total
(BaseWorkerMixin pid=38238) * Remote cpu time: 53.86ms min, 257.69ms max, 133.78ms mean, 13.38s total
(BaseWorkerMixin pid=38238) * Peak heap memory usage (MiB): 1465139200000.0 min, 2456166400000.0 max, 2040879800320 mean
(BaseWorkerMixin pid=38238) * Output num rows: 500 min, 500 max, 500 mean, 50000 total
(BaseWorkerMixin pid=38238) * Output size bytes: 102400128 min, 102400128 max, 102400128 mean, 10240012800 total
(BaseWorkerMixin pid=38238) * Tasks per node: 100 min, 100 max, 100 mean; 1 nodes used
(BaseWorkerMixin pid=38238) 
(BaseWorkerMixin pid=38238) Dataset iterator time breakdown:
(BaseWorkerMixin pid=38238) * In ray.wait(): 18.81ms
(BaseWorkerMixin pid=38238) * In ray.get(): 6.22s
(BaseWorkerMixin pid=38238) * In format_batch(): 27.67ms
(BaseWorkerMixin pid=38238) * In user code: 136.39ms
(BaseWorkerMixin pid=38238) * Total time: 6.41s
(BaseWorkerMixin pid=38238) 
```
```
+--------------------------+------------+-----------------+--------+------------------+--------------+----------------+---------------+
| Trial name               | status     | loc             |   iter |   total time (s) |   bytes_read |   batches_read |   epochs_read |
|--------------------------+------------+-----------------+--------+------------------+--------------+----------------+---------------|
| DummyTrainer_576dd_00000 | TERMINATED | 127.0.0.1:38120 |    100 |          54.7721 |  10240000000 |            100 |             1 |
+--------------------------+------------+-----------------+--------+------------------+--------------+----------------+---------------+
```

**Review by jianoaix — COMMENTED** [review]
LG!
Can you also update the https://docs.ray.io/en/master/ray-air/check-ingest.html?

**Comment by matthewdeng** [comment]
> LG! Can you also update the https://docs.ray.io/en/master/ray-air/check-ingest.html?

Anything in particular to update? I've updated the code snippet used in this page ([doc/source/ray-air/doc_code/air_ingest.py](https://github.com/ray-project/ray/pull/26175/files#diff-6737ccb9077ea688668ef8e24e921325fcca81344eb5769b44dc770d8f2a3ffb)).

**Review by jianoaix — APPROVED** [review]

**Pull request merged**